clash-cores/src/Clash/Cores/Sgmii/PcsTransmit/OrderedSet.hs

```diff
@@ -0,0 +1,229 @@
+  | XmitData {_xmit :: Xmit, _xmitChange :: Bool}
+  | StartOfPacket {_xmit :: Xmit, _xmitChange :: Bool}
+  | TxData {_xmit :: Xmit, _xmitChange :: Bool}
+  | EndOfPacketNoExt {_xmit :: Xmit, _xmitChange :: Bool}
+  | Epd2NoExt {_xmit :: Xmit, _xmitChange :: Bool}
+  | Epd3 {_xmit :: Xmit, _xmitChange :: Bool}
+  | EndOfPacketExt {_xmit :: Xmit, _xmitChange :: Bool}
+  | ExtendBy1 {_xmit :: Xmit, _xmitChange :: Bool}
+  | CarrierExtend {_xmit :: Xmit, _xmitChange :: Bool}
+  | StartError {_xmit :: Xmit, _xmitChange :: Bool}
+  | TxDataError {_xmit :: Xmit, _xmitChange :: Bool}
+  deriving (Generic, NFDataX, Show)
+
+-- | State transitions from @TX_TEST_XMIT@ from Figure 36-5, which need to be
+--   set in all parent states of @TX_TEST_XMIT@ as this state itself is not
+--   implemented as it does not transmit a code group
+txTestXmit ::
+  Bool -> Bool -> Xmit -> Even -> Bool -> Bool -> Maybe OrderedSetState
+txTestXmit txEn txEr xmit txEven tx xmitChange
+  | not (xmitChange && tx && txEven == Odd) = Nothing
+  | xmit == Conf = Just (Configuration xmit False)
+  | xmit == Idle = Just (IdleS xmit False)
+  | xmit == Data && txEn = Just (IdleS xmit False)
+  | xmit == Data && txEr = Just (IdleS xmit False)
+  | otherwise = Just (XmitData xmit False)
+
+-- | Void function that is used to check whether @/V/@ needs to be propagated
+--   based on the values of the input pins
+void :: OrderedSet -> Bool -> Bool -> BitVector 8 -> OrderedSet
+void txOSet txEn txEr dw
+  | not txEn && txEr && dw /= 0b00001111 = OSetV
+  | txEn && txEr = OSetV
+  | otherwise = txOSet
+
+-- | Function to update the current values for 'Xmit' and @xmitChange@
+xmitUpdate :: OrderedSetState -> Maybe Xmit -> (Xmit, Bool)
+xmitUpdate s xmit = (xmit', xmitChange)
+ where
+  xmit' = fromMaybe (_xmit s) xmit
+  xmitChange = (xmit' /= _xmit s) || _xmitChange s
+
+-- | State transition function for the states as defined in IEEE 802.3 Clause
+--   36, specifically Figure 36-5. This function receives the input values and
+--   generates an ordered set to be transmitted by the code group process.
+--
+--   __N.B.__: This function does not implement the optional EEE
+--   (Energy-Efficient Ethernet) capability.
+orderedSetT ::
+  -- | State variable
+  OrderedSetState ->
+  -- | The new input values, partly from the outside world, and partly from
+  --   'Sgmii.autoNeg' and 'PcsTransmit.codeGroupT'
+  (Bool, Bool, BitVector 8, Maybe Xmit, Even, Bool) ->
+  -- | The new state and the new output values
+  (OrderedSetState, (OrderedSetState, OrderedSet))
+orderedSetT s@Configuration{} (txEn, txEr, _, xmit, txEven, tx) =
+  (nextState, out)
+ where
+  nextState = fromMaybe (Configuration xmit' xmitChange) s'
+
+  (xmit', xmitChange) = xmitUpdate s xmit
+  s' = txTestXmit txEn txEr xmit' txEven tx xmitChange
+  out = (s, OSetC)
+orderedSetT s@IdleS{} (txEn, txEr, _, xmit, txEven, tx) = (nextState, out)
+ where
+  nextState
+    | isJust s' = fromJust s'
```

Comment:
```suggestion
    | Just x <- s' = x
```
And then similarly for other places where you've used this `isJust`/`fromJust` combinatrion.